### DISCUSSION TITLE
Unicode issues with ODA File Converter 24.1x

### DISCUSSION THREAD

**Discussion by xbdr419** [mozman/ezdxf#1018] [Issues] [answered]
![image](https://github.com/mozman/ezdxf/assets/19528714/fe8c817d-a01a-4dcb-a778-26adcab7c965)

Codes:
```
from ezdxf.addons import odafc
doc = odafc.readfile(r'test.dwg')
msp = doc.modelspace()
for e in msp:
    if e.dxftype() == 'MTEXT':
        if 'B133002587' in e.dxf.text:
            print(e.dxf.text)
```
Result:
![image](https://github.com/mozman/ezdxf/assets/19528714/24ba37fc-00e7-462f-b7d2-7515f1b6cd3f)

File:
[test.zip](https://github.com/mozman/ezdxf/files/140596981/test.zip)

Hello， When I try to get the text in the dwg, But the returned data is similar to Unicode，How can I get the normal text?
Ths !

**Comment by mozman** [comment]
The conversion from AutoCAD Unicode encoding to regular Unicode should happen automatically.

You can try to convert the strings manually:

```python
from ezdxf.lldxf.encoding import decode_dxf_unicode

# snip x-x-x-x-x

print(decode_dxf_unicode(e.dxf.text))
```
It appears that `\U+00BF` is an ASCII codepage encoding and, as far as I know, defeats the purpose of Unicode encoding. This is the output on my system:

```
In [3]: decode_dxf_unicode(r"\U+00BF")
Out[3]: '¿'
```

**Reply by mozman** [reply]
There seems to be something wrong with the `odafc` add-on. When I convert your test file to DXF by the desktop application `ODA File Converter 23.3.0` to DXF 2004 (the same version as the DWG) the converted text looks correct:

![image](https://github.com/mozman/ezdxf/assets/4230966/783990d6-4b2d-444f-920d-62a89397b1d8)

![image](https://github.com/mozman/ezdxf/assets/4230966/85cc5f2d-ead6-44da-aa9b-897daba0a4d9)

**Reply by mozman** [reply]
There is something wrong with your environment/installation:

```python
from ezdxf.addons import odafc


def main():
    doc = odafc.readfile(r"C:\Users\mozman\Desktop\Now\ezdxf\1018\test\test.dwg")
    print(f"loaded document has DXF version: {doc.acad_release}")
    mtext = doc.entitydb.get("265AA")
    print(mtext.dxf.text)


if __name__ == "__main__":
    main()
```

Prints the correct text:
```
loaded document has DXF version: R2004
勘甲级{\H1.1667x;B133002587}
```

**Comment by xbdr419** [comment, marked as answer]
tks，The problem has been resolved.
My ODA File Converter version is **24.11**，and I try to update lastes  **24.12**，bug the problem still exists.
now ，I  installed the same version (ODA File Converter 23.3.0) as you  ，I have obtained the correct text
so the fundamental reason is still in ODA.

**Reply by mozman** [reply]
Version 23.12 on Windows is OK and doesn't have the bug.

## Download Links
Windows: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_vc16_amd64dll_23.12.msi
DEB: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_lnxX64_8.3dll_23.12.deb
RPM: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_lnxX64_8.3dll_23.12.rpm
AppImage: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_lnxX64_8.3dll_23.12.AppImage
Mac OS X DMG: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_macOsX_x64_11.0dll_23.12.dmg
Mac OS X PKG: https://download.opendesign.com/guestfiles/Demo/ODAFileConverter_QT5_macOsX_x64_11.0dll_23.12.pkg

**Reply by mozman** [reply]
I revisited this issue and it is solved in the latest version of ODAFileConverter v25.4.0.

**Comment by mozman** [comment]
On Linux this issue is not present. The Linux AppImage version 24.12 (and any earlier version) converts the text to MIF encoding:
![browse-odafc-24-12](https://github.com/mozman/ezdxf/assets/4230966/cccd4267-5ae4-49ed-9091-3e4372895918)

Fortunately, support for MIF encoding was added to `ezdxf` in version v1.1.1 (only via `recover` mode!) and after installing the font "simsun.ttc" the text is rendered correctly:

![view-odafc-24-12](https://github.com/mozman/ezdxf/assets/4230966/2b228ba8-3bbc-4907-9d02-46bf57d9fad1)

**Reply by mozman** [reply]
The tofu-rendering in section F/1 is caused by a missing font that has been replaced with the default font. On Windows the default font is `ArialUni.ttf`, a font with a very extensive character support, on Linux it is `DeJaVuSans.ttf` with a much lower character support, hence tofu-rendering.
